core-container/entry.sh

```diff
@@ -23,145 +23,122 @@ function echoerr() {
-    kubectl create secret generic -n ${NAMESPACE} kcpsec --from-file=kubeconfig="kcpsecret"    
-    rm kcpsecret
+function get_kcp_kubeconfig() {
+    while ! KUBECONFIG=$host_kubeconfig kubectl get secret kubestellar; do
+        echo "Waiting for kubestellar secret."
+        sleep 1
+    done
+    kcp_kubeconfig_dir="/home/kubestellar/.kcp"
+    kcp_kubeconfig="${kcp_kubeconfig_dir}/admin.kubeconfig"
+    external_kcp_kubeconfig="${kcp_kubeconfig_dir}/external.kubeconfig"
+    mkdir -p $kcp_kubeconfig_dir
+    echo "Copying the kubeconfigs from kubestellar seret into ${kcp_kubeconfig} and ${external_kcp_kubeconfig}..."
+    kubectl --kubeconfig $host_kubeconfig get secrets kubestellar -o jsonpath='{.data.admin\.kubeconfig}' | base64 --decode > $kcp_kubeconfig
+    kubectl --kubeconfig $host_kubeconfig get secrets kubestellar -o jsonpath='{.data.external\.kubeconfig}' | base64 --decode > $external_kcp_kubeconfig
+}
 
-    echo "Delete the kcp provider object if it already exists, and then create it."
-    if ! kubectl delete spaceproviderdesc pkcp ; then
-        echo "Nothing to delete."
-    fi
-    kubectl create -f - <<EOF
+function create_kcp_provider_object() {
+    get_kcp_kubeconfig     # kcp is created in a seperate container 
+    kubectl --kubeconfig $SPACE_MANAGER_KUBECONFIG delete secret -n ${NAMESPACE} kcpsec > /dev/null 2>&1 || true
+    kubectl --kubeconfig $SPACE_MANAGER_KUBECONFIG create secret generic -n ${NAMESPACE} kcpsec --from-file=kubeconfig=$kcp_kubeconfig --from-file=external=$external_kcp_kubeconfig
+    kubectl --kubeconfig $SPACE_MANAGER_KUBECONFIG apply -f - <<EOF
 apiVersion: space.kubestellar.io/v1alpha1
 kind: SpaceProviderDesc
 metadata:
-  name: pkcp
+  name: $PROVIDER_NAME
 spec:
   ProviderType: "kcp"
   SpacePrefixForDiscovery: "ks-"
   secretRef:
     namespace: ${NAMESPACE}
     name: kcpsec
 EOF
-
-    if [ "$SPACE_PROVIDER" == "kubeflex" ]; then
-        echo "Waiting for the kubeflex provider to be ready... this may take a while."
-        (
-            until [ "$(kubectl get pods -A | grep kubeflex-controller-manager | grep Running)" != "" ]; do
-                sleep 10
-            done
-        )
+    echo "Waiting for spaceproviderdesc to reach the Ready phase."
+    kubectl --kubeconfig ${SPACE_MANAGER_KUBECONFIG} wait --for=jsonpath='{.status.Phase}'=Ready spaceproviderdesc $PROVIDER_NAME
+}
  
-        echo "Delete the kubeflex provider object if it already exists, and then create it."
-        if ! kubectl delete spaceproviderdesc pkflex ; then
-            echo "Nothing to delete."
-        fi
-        kubectl create -f - <<EOF
+function create_kubeflex_provider_object() {
+    echo "Waiting for the kubeflex provider to be ready... this may take a while."
+    (
+        until [ "$(kubectl --kubeconfig $host_kubeconfig get pods -A | grep kubeflex-controller-manager | grep Running)" != "" ]; do
+            sleep 10
+        done
+    )
```

Comment:
What goes wrong if these lines are deleted?

Reply:
In theory we should be good without these lines.  The rest of this function would finish successfully.  However, we need to verify code further up the stack which then creates and uses spaces waits for the spaces to reach the ready state (which it should already).  I will open an issue for confirming this in 1286.